File: removeNthNodeFromEndOfList/c++/Remove_nth_node_from_end_of_list.cpp
//
//  Remove_nth_node_from_end_of_list.cpp
//
//
//  Created by Wen-Ting Wang on 2018/10/31.
//

#include <iostream>
using namespace std;

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

/* another way
class Solution{
public:
    ListNode* removeNthFromEnd(ListNode* head, int n){
        
        if(!head->next) return NULL;
        int length = 0;
        ListNode *tmp = root;
        
        while(tmp){
            tmp = tmp->next;
            length ++;
        }
        
        if(length == n) return head->next;
        else{
            ListNode* current = head;
            for(int i = 0; i < length - n - 1; ++i)
                current = current->next;
            current->next = current->next->next;
            return head;
        }
    }
    
}
*/

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        if(!head->next) return NULL;
        
        ListNode *keep = head, *move = head;
        
        for(int i = 0; i < n; ++i) move = move->next;
        if(!move) return head->next;
        
        while(move->next){
            move = move->next;
            keep = keep->next;
        }
        keep->next = keep->next->next;
        
        return head;
    }
};


//    作者：SunnyYoona
//    来源：CSDN
//    原文：https://blog.csdn.net/sunnyyoona/article/details/42462457
ListNode* createList(int *A, int n){
    
    if(n <= 0) return NULL;
    
    ListNode *head = new ListNode(A[0]), *temp = head;
    
    for(int i = 1; i < n; i++){
        ListNode *input = new ListNode(A[i]);
        temp->next = input;
        temp = temp->next;
    }
    
    return head;
    
}




int main() {
    
    int A[] = {1,2,4,7,9};
    
    
    ListNode* temp = createList(A,5);
    
    Solution ans;
    ListNode *head = ans.removeNthFromEnd(temp, 1);
    
    ListNode *p = head;
    while(p){
        cout<<p->val<<" ";
        p = p->next;
    }
    cout<<endl;
    
    return 0;
}

File: problems/removeNthNodeFromEndOfList/python/Remove_nth_node_from_end_of_list.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
  def __init__(self, x):
       self.val = x
       self.next = None

class Solution:
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        if(head.next == None):
            return None
        tmp = head;

        for i in range(n):
            tmp = tmp.next
        if(tmp == None):
            return head.next
        else:
            keep = head
            while(tmp.next != None):
                keep = keep.next
                tmp = tmp.next

            keep.next = keep.next.next
            return head



class Solution2:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        total = 0

        current = head
        while current:
            current = current.next
            total += 1
        if total == n:
            return head.next

        previous = None
        current = head
        position = total - n + 1

        while position > 1:
            previous = current
            current = current.next
            position -= 1

        if current and current.next:
            previous.next = previous.next.next
        else:
            previous.next = None
        return head


class Solution3:
    """
    - Notes: two-pointer
    """
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        fast = head
        slow = head

        while n > 0:
            fast = fast.next
            n -= 1

        if not fast:
            return head.next

        while fast and fast.next:
            fast = fast.next
            slow = slow.next

        slow.next = slow.next.next
        return head



if __name__ == "__main__":
  data = ListNode(1)
  data.next = ListNode(2)
  data.next.next = ListNode(3)
  data.next.next.next = ListNode(4)
  data.next.next.next.next = ListNode(5)

  ans = Solution()

  ansList = ans.removeNthFromEnd(data, 2)
  for i in range(4):
    print("%d \t"% ansList.val)
    ansList = ansList.next


  data = ListNode(1)
  data.next = ListNode(2)
  data.next.next = ListNode(3)
  data.next.next.next = ListNode(4)
  data.next.next.next.next = ListNode(5)

  ans = Solution2()

  ansList = ans.removeNthFromEnd(data, 2)
  for i in range(4):
    print("%d \t"% ansList.val)
    ansList = ansList.next


  data = ListNode(1)
  data.next = ListNode(2)
  data.next.next = ListNode(3)
  data.next.next.next = ListNode(4)
  data.next.next.next.next = ListNode(5)

  ans = Solution3()

  ansList = ans.removeNthFromEnd(data, 2)
  for i in range(4):
    print("%d \t"% ansList.val)
    ansList = ansList.next
